frontend-angularjs/app/scripts/controllers/linesCtrl.js

```diff
@@ -1,3 +1,306 @@
+      }
+    });
+  };
+
+  $scope.openEditLineModal = function(line) {
+    $uibModal.open({
+      templateUrl: 'lineModal.html',
+      controller: 'EditLineCtrl',
+      resolve: {
+        lineData: function () {
+          return angular.copy(line);
+        }
+      }
+    }).result.then(function (editedLine) {
+      if (editedLine.error) {
+        $scope.addAlert('Error happened: \'' + editedLine.error + ' \'', 'danger');
+      } else {
+        $scope.loadPage($scope.currentPageIndex(), function () {
+          $scope.addAlert('Line \'' + editedLine.name + ' \' has been successfully saved!', 'success');
+        });
+      }
+    });
+  };
+});
+
+angular.module('hermesApp').controller('AbstractLineModalCtrl', function ($q, $scope, $timeout, $http, $uibModalInstance, env) {
+
+  $scope.persistRoute = function (code, vehicleType, stations, callback, url) {
+    const reqData = {
+      code: code,
+      vehicleType: vehicleType,
+      stations: stations
+    };
+
+    if (!url) {
+      return $http.post(env.backendBaseUrl + "/routes", reqData)
+                  .then(function(result) {callback(result);},
+                        function(error) {$scope.addAlert(error);});
+    }
+    return $http.patch(url, reqData)
+                .then(function(result) {callback(result);},
+                      function(error) {$scope.addAlert(error);});
+  };
+
+  $scope.persistLine = function (code, name, agency, infoUrl, inboundRoute, outboundRoute, callback, url) {
+    const reqData = {
+      code: code,
+      name: name,
+      agency: agency,
+      url: infoUrl,
+      inboundRoute: inboundRoute,
+      outboundRoute: outboundRoute
+    };
+
+    if (!url) {
+      $http.post(env.backendBaseUrl + "/lines", reqData)
+        .then(function (response) {
+          if (typeof callback === 'function')
+            callback({
+              name: response.data.name,
+              href: response.data._links.line.href
+            });
+        }, function (error) {
+          if (typeof callback === 'function')
+            callback({
+              error: error
+            });
+        });
+    } else {
+      $http.patch(url, reqData)
+        .then(function (response) {
+          if (typeof callback === 'function')
+            callback({
+              name: response.data.name,
+              href: response.data._links.line.href
+            });
+        }, function (error) {
+          if (typeof callback === 'function')
+            callback({
+              error: error
+            });
+        });
+    }
+  };
+
+  $scope.initModal = function () {
+    $.material.init();
+
+    // Init sortable
+    $('.sortable-list').each(function(i, obj) {
+     Sortable.create(obj, {handle: '.list-group-sortable', draggable: ".list-group-sortable-item"});
+    });
+  };
+
+  $scope.fetchStationsContaining = function (name, filterList) {
+    return $http.get(env.backendBaseUrl + "/stations/search/findByNameContainingIgnoreCase?name=" + name)
```

Comment:
In my opinion, when you do filtering you should also take into accound vehicle type chosen on ui. This way you won't have stations with vehicle types incompatible with route vehicle type